File: packages/paperclip/src/parser/virt.rs

pub struct Element<'a> {
  pub tag_name: &'a str,
  pub attributes: Vec<Attribute<'a>>,
  pub children: Vec<Node<'a>>
}

pub struct Attribute<'a> {
  pub name: &'a str,
  pub value: Option<&'a str>
}

pub struct Text<'a> {
  pub value: &'a str
}

pub enum Node<'a> {
  Element(Element<'a>),
  Text(Text<'a>)
}

File: packages/paperclip/src/virt.ts
export enum VirtualNodeType {
  Element = "Element",
  Text = "Text",
  Fragment = "Fragment"
}

type VirtualBaseNode<TType extends VirtualNodeType> = {
  type: TType;
};

export type VirtualAttribute = {
  name: string;
  value: string;
};
export type VirtualElement = {
  attributes: VirtualAttribute[];
  children: VirtualNode;
} & VirtualBaseNode<VirtualNodeType.Element>;

export type VirtualText = {
  value: string;
} & VirtualBaseNode<VirtualNodeType.Element>;

export type VirtualNode = VirtualElement | VirtualText;
